link-grammar/fast-match.c

```diff
@@ -203,6 +204,59 @@ static Match_node * add_to_left_table_list(Match_node * m, Match_node * l)
 }
 
 /**
+ * Validate that the uppercase part of two connectors are the same.
+ * Return true if so, false otherwise.
+ * FIXME: Use connector enumeration.
+ */
+static bool con_uc_eq(const Connector *c1, const Connector *c2)
+{
+
+	if (string_set_cmp(c1->string, c2->string)) return true;
+	if (c1->hash != c2->hash) return false;
+	if (c1->uc_length != c1->uc_length) return false;
+
+	const char *uc1 = &c1->string[c1->uc_start];
+	const char *uc2 = &c2->string[c2->uc_start];
+	/* We arrive here for less than 50% of the cases for "en" and
+	 * less then 20% of the cases for "ru", and the following
+	 * condition is always true.
+	 * How come 2 different uc strings never have the same hash value? */
```

Comment:
You have a terrible hashing algo.

There are about 100 or so different UC connectors in English. With a good hash, that should fit just fine into a single byte (256 points).  Should fit trivially into 32 bits, as none of the connectors are longer than 3 UC letters.  Most are 1 or 2 letters, and only a couple are 3 UC letters long, so if your hash can't fit them into 65536 slots, you've got a terrible hash.  I mean, you could just hand-assemble them ..
```
int16 mask (char c) { return  (c-0x40)  & 0x20; }
int16 hash = mask(s[0]) | (mask(s[1])<<5) | mask(s[2]) << 10;
```
the above should give you a perfect hash for the English dicts, since all connectors are 3 UC letters long or less. thus the strcmp would never-ever happen.